api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java

```diff
@@ -0,0 +1,249 @@
+              .description(conceptSet.getDescription())
+              .id(conceptSet.getConceptSetId())
+              .name(conceptSet.getName())
+              .domain(conceptSet.getDomainEnum());
+          if (conceptSet.getCreator() != null) {
+            result.creator(conceptSet.getCreator().getEmail());
+          }
+          return result;
+        }
+      };
+
+  private static final Function<ConceptSet, org.pmiops.workbench.db.model.ConceptSet> FROM_CLIENT_CONCEPT_SET =
+      new Function<ConceptSet, org.pmiops.workbench.db.model.ConceptSet>() {
+        @Override
+        public org.pmiops.workbench.db.model.ConceptSet apply(ConceptSet conceptSet) {
+          org.pmiops.workbench.db.model.ConceptSet dbConceptSet =
+              new org.pmiops.workbench.db.model.ConceptSet();
+          dbConceptSet.setDomainEnum(conceptSet.getDomain());
+          dbConceptSet.setDescription(conceptSet.getDescription());
+          dbConceptSet.setName(conceptSet.getName());
+          return dbConceptSet;
+        }
+      };
+
+  private static final Ordering<Concept> CONCEPT_NAME_ORDERING =
+      Ordering.from(String.CASE_INSENSITIVE_ORDER).onResultOf(Concept::getConceptName);
+
+  @Autowired
+  ConceptSetsController(WorkspaceService workspaceService, ConceptSetDao conceptSetDao,
+      ConceptDao conceptDao, Provider<User> userProvider, Clock clock) {
+    this.workspaceService = workspaceService;
+    this.conceptSetDao = conceptSetDao;
+    this.conceptDao = conceptDao;
+    this.userProvider = userProvider;
+    this.clock = clock;
+  }
+
+  @Override
+  public ResponseEntity<ConceptSet> createConceptSet(String workspaceNamespace, String workspaceId,
+      ConceptSet conceptSet) {
+    Workspace workspace = workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(
+        workspaceNamespace, workspaceId, WorkspaceAccessLevel.WRITER);
+    org.pmiops.workbench.db.model.ConceptSet dbConceptSet = FROM_CLIENT_CONCEPT_SET.apply(conceptSet);
+    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+    dbConceptSet.setCreator(userProvider.get());
+    dbConceptSet.setWorkspaceId(workspace.getWorkspaceId());
+    dbConceptSet.setCreationTime(now);
+    dbConceptSet.setLastModifiedTime(now);
+    dbConceptSet.setVersion(1);
+    try {
+      dbConceptSet = conceptSetDao.save(dbConceptSet);
+      // TODO: add recent resource entry for concept sets?
+    } catch (DataIntegrityViolationException e) {
+      throw new BadRequestException(String.format(
+          "Concept set \"/%s/%s/%d\" already exists.",
+          workspaceNamespace, workspaceId, dbConceptSet.getName()));
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  private ConceptSet toClientConceptSet(org.pmiops.workbench.db.model.ConceptSet conceptSet) {
+    ConceptSet result = TO_CLIENT_CONCEPT_SET.apply(conceptSet);
+    if (!conceptSet.getConceptIds().isEmpty()) {
```

Comment:
Is there a reason not to do this in `TO_CLIENT_CONCEPT_SET`?

Reply:
TO_CLIENT_CONCEPT_SET doesn't do any RPCs.... which is appropriate. We don't want to issue N RPCs when dealing with N concept sets.

In the list case, we use it to convert the concept sets without the concepts underneath, and then separately fetch the concepts for all N concept sets in one go from the DB. That way we're still doing a constant number of database lookups.

Reply:
Got it. That sounds good! 👍